### PR TITLE
TDL: Removed retry for 400 and added reauthentication for Connection …

### PR DESCRIPTION
**Pull request by dbshah1212**
# Description of change
Removed retry for 400 and added retry and reauthentication for 401,429, 502,503,504. 
Made updated_at of product stream available as product stream is Full_Table
Added Annotation for High-level main method exception. 

# Manual QA steps
 - Raised some errors except 400,403 and 404 and checked that reauthentication is happening. 
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

**Review by savan-chovatiya — APPROVED** [review]

File: tap_yotpo/http.py [hunk 1]
```diff
@@ -147,38 +148,36 @@ def authenticate(self):
...
+            response_json = {}
+        if response.status_code in ERROR_CODE_EXCEPTION_MAPPING.keys():
+            message = "HTTP-error-code: {}, Error: {}".format(
+                response.status_code,
+                response_json.get("status", ERROR_CODE_EXCEPTION_MAPPING.get(
+                    response.status_code, {})).get("message", "Unknown Error")
+            )
+            exc = ERROR_CODE_EXCEPTION_MAPPING.get(
+                response.status_code, {}).get("raise_exception", YotpoError)
+            if not authentication_call and response.status_code in [401, 502]:
```
Comment (karanpanchal-crest) [inline tap_yotpo/http.py:175]: I think we should retry for error codes except 400, 403, and 404 from the list of status codes supported by Yotpo.
@dbshah1212 Please negate this statement to skip retry mechanism for above mentioned error codes

File: tap_yotpo/http.py [hunk 2]
```diff
@@ -147,38 +148,36 @@ def authenticate(self):
...
-                exc = ERROR_CODE_EXCEPTION_MAPPING.get(
-                    error_code, {}).get("raise_exception", YotpoError)
-                raise exc(message, resp) from None
-
-            except (ValueError, TypeError):
-                raise YotpoError(error) from None
+            response_json = response.json()
+        except Exception:
+            response_json = {}
+        if response.status_code in ERROR_CODE_EXCEPTION_MAPPING.keys():
```
Comment (karanpanchal-crest) [inline tap_yotpo/http.py:167]: `if response.status_code in ERROR_CODE_EXCEPTION_MAPPING.keys() or response.status_code != 200:`
Please re-structure the code as the above condition

Comment (dbshah1212) [reply tap_yotpo/http.py:167]: Only response.status_code != 200 is sufficed so removing other conditions as well.

File: tap_yotpo/http.py (hunk 1 above)
Comment (dbshah1212) [reply tap_yotpo/http.py:175]: Done

File: tap_yotpo/http.py [hunk 3]
```diff
@@ -147,38 +148,31 @@ def authenticate(self):
...
+            response_json = {}
+        if response.status_code != 200:
+            message = "HTTP-error-code: {}, Error: {}".format(
+                response.status_code,
+                response_json.get("status", ERROR_CODE_EXCEPTION_MAPPING.get(
+                    response.status_code, {})).get("message", "Unknown Error")
+            )
+            exc = ERROR_CODE_EXCEPTION_MAPPING.get(
+                response.status_code, {}).get("raise_exception", YotpoError)
+            # Reauthenticating if it isn't raise while authentication and for all error except some known error
```
Comment (karanpanchal-crest) [inline tap_yotpo/http.py:175]: @dbshah1212 Please update the comment as follows:
"Re-authenticating if got an authentication error while collecting stream data and response does not have Bad request (400), Forbidden (403) and Not found (404)

Comment (dbshah1212) [reply tap_yotpo/http.py]: Updated

**Review by karanpanchal-crest — APPROVED** [review]

**Review by KAllan357 — APPROVED** [review]

**Pull request merged**